lib/scala/semver/src/main/java/org/enso/semver/SemVer.java

```diff
@@ -0,0 +1,145 @@
+        this.prefix = Optional.empty();
+    }
+
+    public static Try<SemVer> parse(String version) {
+        var match = regex.matcher(version);
+        if (match.find()) {
+            Optional<String> prefix = match.groupCount() > 0 && match.group(1) != null ? Optional.of(match.group(1)) : Optional.empty();
+            Optional<String> preRelease = match.groupCount() == 5 && match.group(5) != null ? Optional.of(match.group(5).substring(1)) : Optional.empty();
+            return Success.apply(new SemVer(
+                    Integer.valueOf(match.group(2)),
+                    Integer.valueOf(match.group(3)),
+                    Integer.valueOf(match.group(4)),
+                    preRelease,
+                    prefix));
+
+        } else {
+            return Failure.apply(new RuntimeException("Invalid version " + version));
+        }
+    }
+
+    public static SemVer of(int major) {
+        return new SemVer(major, 0, 0);
+    }
+
+    public static SemVer of(int major, int minor) {
+        return new SemVer(major, minor, 0);
+    }
+
+    public static SemVer of(int major, int minor, int patch) {
+        return new SemVer(major, minor, patch);
+    }
+
+    public static SemVer of(int major, int minor, int patch, String preReleaseVersion) {
+        return new SemVer(major, minor, patch, Optional.of(preReleaseVersion));
+    }
+
+    public boolean isGreaterThanOrEqual(SemVer v) {
+        return major > v.major || major == v.major && (
+                minor > v.minor || minor == v.minor && patch >= v.patch);
+    }
+
+    public boolean isGreaterThan(SemVer v) {
+        return major > v.major || major == v.major && (
+                minor > v.minor || minor == v.minor && patch > v.patch);
+    }
+
+    public boolean isLessThanOrEqual(SemVer v) {
+        return major < v.major || major == v.major && (
+                minor < v.minor || minor == v.minor && patch <= v.patch);
+    }
+
+    public boolean isLessThan(SemVer v) {
+        return major < v.major || major == v.major && (
+                minor < v.minor || minor == v.minor && patch < v.patch);
+    }
+
+    public Optional<String> preReleaseVersion() {
+        return preRelease;
+    }
+
+    public int compareTo(SemVer v) {
+        if (major < v.major) return -1;
+        else if (major > v.major) return 1;
+        else {
+            if (minor < v.minor) return -1;
+            else if (minor > v.minor) return 1;
+            else {
+                if (patch < v.patch) return -1;
+                else if (patch > v.patch) return 1;
+                else {
+                    if (preRelease.isEmpty() && v.preRelease.isEmpty()) return 0;
+                    if (preRelease.isEmpty()) return 1;
+                    else if (v.preRelease.isEmpty()) return -1;
+                    else return (preRelease.equals(v.preRelease) ? 0 : preRelease.get().compareTo(v.preRelease.get()));
+                }
+            }
+        }
+    }
```

Comment:
Oh, I can see that here you are implementing the ordering correctly, taking into account the prerelease logic - good.

Well - why do we then have 2 implementations of ordering that are inconsistent? I think the `isLessThanOrEqual` and co. should be calling into `compareTo`, so that we get a consistent implementation. 

If not, we should have tests that verify the consistency. But IMO we should just delegate to `compareTo` and DRY, I see no point in duplicating this logic.

Reply:
Yeah, apparently our tests picked that up, hence the difference. I will see if I can integrate that.